lib/src/kbasesearchengine/main/NarrativeInfoDecorator.java

```diff
@@ -105,35 +115,49 @@ public GetObjectsOutput getObjects(final GetObjectsInput params, final String us
-                    String, Map<String,String>> wsInfo;
-            
-            try {
-                wsInfo = weh.getWorkspaceInfo(workspaceId);
-            } catch (IOException e) {
-                throw new IOException("Failed retrieving workspace info: " + e.getMessage(), e);
-            } catch (JsonClientException e) {
-                throw new JsonClientException("Failed retrieving workspace info: "
-                        + e.getMessage(), e);
-            }
-
-            final long timeMilli = WorkspaceEventHandler.parseDateToEpochMillis(wsInfo.getE4());
             final Tuple5<String, Long, Long, String, String> tempNarrInfo =
-                        new Tuple5<String, Long, Long, String, String>()
-                    .withE3(timeMilli)      // modification time
-                    .withE4(wsInfo.getE3()) // workspace user name
-                    .withE5(null);          // TODO workspace user, real name
-
-            final Map<String, String> wsInfoMeta = wsInfo.getE9();
-
-            if (wsInfoMeta.containsKey("narrative") &&
-                    wsInfoMeta.containsKey("narrative_nice_name")) {
-                tempNarrInfo.withE1(wsInfoMeta.get("narrative_nice_name"))
-                        .withE2(Long.parseLong(wsInfoMeta.get("narrative")));
-            }
+                    getNarrativeInfo(workspaceId);
+            userNames.add(tempNarrInfo.getE4());
             retVal.put(workspaceId, tempNarrInfo);
         }
+        final Map<String, String> displayNames = authClient.getUserDisplayNames(token, userNames);
+        // e5 is the full / display name, e4 is the user name
+        // defaults to the existing name so names from previous decorator layers aren't set to null
+        retVal.values().stream().forEach(t -> t.withE5(
+                displayNames.getOrDefault(t.getE4(), t.getE5())));
         return retVal;
     }
+    
+    private Tuple5<String, Long, Long, String, String> getNarrativeInfo(final long wsid)
+            throws IOException, JsonClientException {
+        final Tuple9 <Long, String, String, String, Long, String, String,
+                String, Map<String,String>> wsInfo;
+
+        try {
+            wsInfo = weh.getWorkspaceInfo(wsid);
+        } catch (IOException e) {
+            throw new IOException("Failed retrieving workspace info: " + e.getMessage(), e);
+        } catch (JsonClientException e) {
+            throw new JsonClientException("Failed retrieving workspace info: "
+                    + e.getMessage(), e);
+        }
+        
+        final long timeMilli = WorkspaceEventHandler.parseDateToEpochMillis(wsInfo.getE4());
+        final Tuple5<String, Long, Long, String, String> tempNarrInfo =
+                    new Tuple5<String, Long, Long, String, String>()
+                .withE3(timeMilli)      // modification time
+                .withE4(wsInfo.getE3()); // workspace user name
+                // e5 gets filled in later
```

Comment:
The real user name is not filled here, in order to combine all the userIds and make one single request to the authClient to get the display names. But since we want to cache the narrativeInfo (wsId, Tuple5<String, Long, Long, String, String>), the display name should also be updated in this method in the returned Tuple5,  so that it can be cached.

Reply:
> But since we want to cache the narrativeInfo (wsId, Tuple5<String, Long, Long, String, String>)

I don't think there's any requirement that the cache needs to have any particular data structure.

Reply:
There isn't any such requirement. But i felt if the narrative info for each wsId needs to be cached, then it would be cleaner, if this method is made to return all the values for the given wsID.

Reply:
We probably don't want to make up to 10000 network calls when 1 would do in the name of cleanliness though. Premature optimization and all that but network calls are one case where it's pretty clear cut that if they can be minimized with little effort (as is the case here) that's what should be done.